crates/biome_rowan/src/ast/batch.rs

```diff
@@ -348,89 +343,189 @@ where
     /// To address this case at step 3, when we pop a new change to apply it, we actually aggregate all changes to the current
     /// parent together. This is done by the heap because we also sort by node and it's range.
     ///
-    pub fn commit(self) -> SyntaxNode<L> {
+    /// Text range and text edit can be collected simultanously. They're directly calculated from the commit changes.
+    pub fn commit_with_text_range_and_edit(
+        self,
+        with_text_range_and_edit: bool,
+    ) -> (SyntaxNode<L>, Option<(TextRange, TextEdit)>) {
         let BatchMutation { root, mut changes } = self;
-        // Fill the heap with the requested changes
 
-        while let Some(item) = changes.pop() {
-            // If parent is None, we reached the root
-            if let Some(current_parent) = item.parent {
+        // Ordered text mutation list sorted by text range
+        let mut text_mutation_list: Vec<(TextRange, Option<String>)> = vec![];
```

Comment:
Is it safe to initialise `text_mutation_list` with `Vec::with_capacity(changes.len())`?

Reply:
Yes, this is a good suggestion! Thanks!

All suggestions are applied in https://github.com/biomejs/biome/pull/2237/commits/b4679d3709f60ff85d06194a70acef6a46c9e499